mrmustard/physics/gaussian_integrals.py

```diff
@@ -259,216 +160,307 @@ def reorder_abc(Abc: tuple, order: Sequence[int]):
+        raise ValueError(
+            f"All batch sizes must match when mode='zip': got ({A1.shape[0]}, {b1.shape[0]}, {c1.shape[0]})"
+            f" and ({A2.shape[0]}, {b2.shape[0]}, {c2.shape[0]})"
+        )
+    if not A1.shape[-1] == A1.shape[-2] == b1.shape[-1]:
+        raise ValueError("A1 must be square and b1 must be compatible with A1.")
+
+    if not A2.shape[-1] == A2.shape[-2] == b2.shape[-1]:
+        raise ValueError("A2 must be square and b2 must be compatible with A2.")
+
+    # 2. get shapes and sizes
+
+    batch1 = A1.shape[0]
+    batch2 = A2.shape[0]
+    poly_shape1 = c1.shape[1:]
+    poly_shape2 = c2.shape[1:]
+    m1 = len(poly_shape1)
+    m2 = len(poly_shape2)
+    n1_plus_m1 = A1.shape[-1]
+    n2_plus_m2 = A2.shape[-1]
+    n1 = n1_plus_m1 - m1
+    n2 = n2_plus_m2 - m2
+
+    # 3. join triples
+
+    c1 = math.reshape(c1, (batch1, -1))
+    c2 = math.reshape(c2, (batch2, -1))
+
+    if mode == "kron":
+        A1 = np.repeat(A1, batch2, axis=0)
+        A2 = np.tile(A2, (batch1, 1, 1))
+        A1Z = np.concatenate([A1, np.zeros((batch1 * batch2, n1_plus_m1, n2_plus_m2))], axis=-1)
+        ZA2 = np.concatenate([np.zeros((batch1 * batch2, n2_plus_m2, n1_plus_m1)), A2], axis=-1)
+        b1 = np.repeat(b1, batch2, axis=0)
+        b2 = np.tile(b2, (batch1, 1))
+        c = math.reshape(
+            math.einsum("ba,dc->bdac", c1, c2),
+            (batch1 * batch2,) + poly_shape1 + poly_shape2,
+        )
+    elif mode == "zip":
+        A1Z = np.concatenate([A1, np.zeros((batch1, n1_plus_m1, n2_plus_m2))], axis=-1)
+        ZA2 = np.concatenate([np.zeros((batch1, n2_plus_m2, n1_plus_m1)), A2], axis=-1)
+        c = math.reshape(math.einsum("ba,bc->bac", c1, c2), (batch1,) + poly_shape1 + poly_shape2)
+
+    A = np.concatenate([A1Z, ZA2], axis=-2)
+    A = np.concatenate(
+        [
+            A[:, :n1, :],
+            A[:, n1_plus_m1 : n1_plus_m1 + n2, :],
+            A[:, n1:n1_plus_m1, :],
+            A[:, n1_plus_m1 + n2 :, :],
+        ],
+        axis=-2,
+    )
+    A = np.concatenate(
+        [
+            A[:, :, :n1],
+            A[:, :, n1_plus_m1 : n1_plus_m1 + n2],
+            A[:, :, n1:n1_plus_m1],
+            A[:, :, n1_plus_m1 + n2 :],
+        ],
+        axis=-1,
+    )
+    b = np.concatenate([b1, b2], axis=-1)
+    b = np.concatenate(
+        [
+            b[:, :n1],
+            b[:, n1_plus_m1 : n1_plus_m1 + n2],
+            b[:, n1:n1_plus_m1],
+            b[:, n1_plus_m1 + n2 :],
+        ],
+        axis=-1,
     )
 
+    # 4. return according to inputs
 
-def complex_gaussian_integral_2(
-    Abc1: tuple, Abc2: tuple, idx1: tuple[int, ...], idx2: tuple[int, ...], measure: float = -1
-):
-    r"""Computes the Gaussian integral of the exponential of a complex quadratic form.
-    The integral is defined as (note that in general we integrate over a subset of 2m dimensions):
+    if not batched1 and not batched2:
+        return A[0], b[0], c[0]
```

Comment:
I dont fully understand why we wouldn't we just always return batched? We really should stick to having all internal methods expect `(A,b,c)` to be batched and return things batched for consistency no? With the exceptions being methods with a `batched` parameter exposed to the user for convenience